core/arch/aarchxx/mangle.c

```diff
@@ -577,50 +577,212 @@ shrink_reg_for_param(reg_id_t regular, opnd_t arg)
 }
 #endif /* !AARCH64 */
 
+/* Return true if opnd is a register, but not XSP, or immediate zero on AArch64. */
+static bool
+opnd_is_reglike(opnd_t opnd)
+{
+    return ((opnd_is_reg(opnd) && opnd_get_reg(opnd) != DR_REG_XSP) ||
+            IF_X64_ELSE((opnd_is_immed_int(opnd) && opnd_get_immed_int(opnd) == 0),
+                        false));
+}
+
 uint
 insert_parameter_preparation(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
                              bool clean_call, uint num_args, opnd_t *args)
 {
-    uint i;
-    instr_t *mark = INSTR_CREATE_label(dcontext);
-    PRE(ilist, instr, mark);
+    uint num_regs = num_args < NUM_REGPARM ? num_args : NUM_REGPARM;
+    signed char regs[NUM_REGPARM];
+    int usecount[NUM_REGPARM];
+    ptr_int_t stack_inc = 0;
+    uint i, j;
 
-    ASSERT(num_args == 0 || args != NULL);
-    /* FIXME i#1551, i#1569: we only support limited number of args for now. */
-    ASSERT_NOT_IMPLEMENTED(num_args <= NUM_REGPARM);
+    /* We expect every arg to be an integer or a full-size register. */
     for (i = 0; i < num_args; i++) {
-        if (opnd_is_immed_int(args[i])) {
-            insert_mov_immed_ptrsz(dcontext, opnd_get_immed_int(args[i]),
-                                   opnd_create_reg(regparms[i]),
-                                   ilist, instr_get_next(mark), NULL, NULL);
-        } else if (opnd_is_reg(args[i])) {
-            opnd_t arg = opnd_create_reg(reg_to_pointer_sized(opnd_get_reg(args[i])));
-            if (opnd_get_reg(arg) == DR_REG_XSP) {
-                instr_t *loc = instr_get_next(mark);
-                PRE(ilist, loc, instr_create_save_to_tls
-                    (dcontext, regparms[i], TLS_REG0_SLOT));
-                insert_get_mcontext_base(dcontext, ilist, loc, regparms[i]);
-                PRE(ilist, loc, instr_create_restore_from_dc_via_reg
+        ASSERT(opnd_is_immed_int((args[i])) ||
+               (opnd_is_reg(args[i]) && reg_get_size(opnd_get_reg(args[i])) == OPSZ_PTR));
+    }
+
+    /* Set up stack arguments that are registers (not SP) or zero (on AArch64). */
+    if (num_args > NUM_REGPARM) {
+        uint n = num_args - NUM_REGPARM;
+        stack_inc = ALIGN_FORWARD(n, 2) * XSP_SZ;
+#ifdef AARCH64
+        for (i = 0; i < n; i += 2) {
+            opnd_t *arg0 = &args[NUM_REGPARM + i];
+            opnd_t *arg1 = &args[NUM_REGPARM + i + 1];
```

Comment:
I thought you were referring to the loop, at first: that's safe because n = num_args - NUM_REGPARM. But &args[NUM_REGPARM + i + 1] should also be fine because you're allowed to take the address of the element just beyond the end of the array.